### PR TITLE
errors: improve rate limit error messages for AI agents

### LINKED ISSUE #2385: Improve rate limit error messages for AI agents
### Summary

When the GitHub API returns a rate limit error, the message surfaced to the AI model as a tool result is either missing the retry duration entirely, or buries it inside a noisy raw HTTP string. This makes it hard for agents to know how long to wait before retrying.

All tool errors flow through `NewGitHubAPIErrorResponse` in `pkg/errors/error.go`, which formats the tool result as:

```
<tool message>: <err.Error()>
```

For a **primary rate limit** (`RateLimitError`), the agent sees:

```
search code: GET https://api.github.com/search/code: 403 API rate limit exceeded for user ID 12345. [rate reset in 47s]
```

The reset time is present but buried inside a raw HTTP method + URL + status string.

For a **secondary (abuse) rate limit** (`AbuseRateLimitError`), the agent sees:

```
search code: GET https://api.github.com/search/code: 403 You have exceeded a secondary rate limit and have been temporarily blocked from content creation. Please retry your request again later.
```

The `Retry-After` duration is **completely absent** — it is populated in the struct from the HTTP header but dropped by `AbuseRateLimitError.Error()` (tracked upstream in google/go-github#4180).

### Suggested fix

In `NewGitHubAPIErrorResponse`, use `errors.As` to detect rate limit error types and return a clean, agent-friendly message:

```go
func NewGitHubAPIErrorResponse(ctx context.Context, message string, resp *github.Response, err error) *mcp.CallToolResult {
    apiErr := newGitHubAPIError(message, resp, err)
    if ctx != nil {
        _, _ = addGitHubAPIErrorToContext(ctx, apiErr)
    }

    var rateLimitErr *github.RateLimitError
    if errors.As(err, &rateLimitErr) {
        retryIn := time.Until(rateLimitErr.Rate.Reset.Time).Round(time.Second)
        return utils.NewToolResultError(fmt.Sprintf(
            "%s: GitHub API rate limit exceeded. Retry after %v.",
            message, retryIn))
    }

    var abuseErr *github.AbuseRateLimitError
    if errors.As(err, &abuseErr) {
        if abuseErr.RetryAfter != nil {
            return utils.NewToolResultError(fmt.Sprintf(
                "%s: GitHub secondary rate limit exceeded. Retry after %v.",
                message, abuseErr.RetryAfter.Round(time.Second)))
        }
        return utils.NewToolResultError(fmt.Sprintf(
            "%s: GitHub secondary rate limit exceeded. Wait before retrying.",
            message))
    }

    return utils.NewToolResultErrorFromErr(message, err)
}
```

This gives the agent clear, actionable output:

```
search code: GitHub API rate limit exceeded. Retry after 47s.
```

```
search code: GitHub secondary rate limit exceeded. Retry after 60s.
```

> **Note:** This fix is independent of google/go-github#4180 — it reads `RetryAfter` directly from the struct rather than relying on `.Error()`. The go-github fix benefits all other callers of the library.

> [!NOTE]
> This issue was drafted with the assistance of GitHub Copilot.

### PR DESCRIPTION
**Pull request by danmoseley**
When the GitHub API returns a rate limit error, agents currently receive raw Go HTTP error strings:

```
search code: GET https://api.github.com/search/code: 403 API rate limit exceeded for user ID 12345. [rate reset in 2m59s]
```

These are hard for agents to reliably parse. Without a clear signal to stop and wait, agents may retry immediately — burning through quota faster — or misread the retry delay and either wait too short (triggering another error) or too long (wasting time).

This PR makes `NewGitHubAPIErrorResponse` intercept `*github.RateLimitError` and `*github.AbuseRateLimitError` and return clean, actionable messages instead:

```
search code: GitHub API rate limit exceeded. Retry after 2m59s.
create issue: GitHub secondary rate limit exceeded. Retry after 47s.
create issue: GitHub secondary rate limit exceeded. Wait before retrying.
```

Edge cases handled:
- Reset time already expired or zero → "Wait before retrying."
- `AbuseRateLimitError.RetryAfter` is nil or zero → "Wait before retrying."
- Wrapped errors (unwrapped via `errors.As`)

**Backward compatibility:** The original error is stored in context via `addGitHubAPIErrorToContext` *before* the rate-limit check, so any middleware reading from context is unaffected — only the text returned to the MCP client changes.

Fixes #2385.

> [!NOTE]
> This PR description was drafted with GitHub Copilot.

### CODE REVIEW

File: pkg/errors/error_test.go [hunk 1]
```diff
@@ -1,462 +1,660 @@
...
+			Rate:     github.Rate{Reset: github.Timestamp{Time: resetTime}},
+			Response: &http.Response{StatusCode: 403},
+			Message:  "API rate limit exceeded",
+		}
+		resp := &github.Response{Response: rateLimitErr.Response}
+
+		// When we create an API error response for a rate limit error
+		result := NewGitHubAPIErrorResponse(ctx, "search code", resp, rateLimitErr)
+		expectedRetryIn := time.Until(resetTime).Round(time.Second)
+
```
[the remark this answers is not in the document]
Comment (danmoseley) [reply pkg/errors/error_test.go]: Good catch — fixed in the follow-up commit. `expectedRetryIn` is now computed before calling `NewGitHubAPIErrorResponse`, and the reset time offsets were increased to 20–30 minutes so a 1-second rounding boundary can never cause a mismatch.

File: pkg/errors/error.go [hunk 2]
```diff
@@ -159,6 +161,30 @@ func NewGitHubAPIErrorResponse(ctx context.Context, message string, resp *github
...
+		return utils.NewToolResultError(fmt.Sprintf(
+			"%s: GitHub API rate limit exceeded. Wait before retrying.", message))
+	}
+
+	var abuseErr *github.AbuseRateLimitError
+	if stderrors.As(err, &abuseErr) {
+		if abuseErr.RetryAfter != nil && *abuseErr.RetryAfter > 0 {
+			return utils.NewToolResultError(fmt.Sprintf(
+				"%s: GitHub secondary rate limit exceeded. Retry after %v.",
+				message, abuseErr.RetryAfter.Round(time.Second)))
```
[the remark this answers is not in the document]
Comment (danmoseley) [reply pkg/errors/error.go]: Thanks for the review. Actually, `abuseErr.RetryAfter.Round(time.Second)` does compile — Go auto-dereferences pointer receivers for value-receiver method calls, so calling `.Round()` on a `*time.Duration` is valid. The code compiled successfully and all 21 tests passed before the push.

Applying the explicit dereference suggestion would introduce a `gocritic` lint violation (`underef: could simplify (*abuseErr.RetryAfter).Round to abuseErr.RetryAfter.Round`), which this repo's `.golangci.yml` enforces. Keeping the original form.

File: pkg/errors/error.go [hunk 3]
```diff
@@ -159,6 +161,30 @@ func NewGitHubAPIErrorResponse(ctx context.Context, message string, resp *github
...
 		_, _ = addGitHubAPIErrorToContext(ctx, apiErr) // Explicitly ignore error for graceful handling
 	}
+
+	var rateLimitErr *github.RateLimitError
+	if stderrors.As(err, &rateLimitErr) {
+		resetTime := rateLimitErr.Rate.Reset.Time
+		if !resetTime.IsZero() && resetTime.After(time.Now()) {
+			retryIn := time.Until(resetTime).Round(time.Second)
+			return utils.NewToolResultError(fmt.Sprintf(
+				"%s: GitHub API rate limit exceeded. Retry after %v.", message, retryIn))
```
[the remark this answers is not in the document]
Comment (danmoseley) [reply pkg/errors/error.go]: Good catch. Fixed in the follow-up commit: 	ime.Until(resetTime) is now computed once and the rounded result is checked >0 before returning the 'Retry after' message. If the rounded value is 0 (e.g. sub-500ms remaining), we fall through to 'Wait before retrying.' instead. A test for this edge case was added.

File: pkg/errors/error.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (danmoseley) [reply pkg/errors/error.go]: Fixed: now round RetryAfter first, then check the rounded value is >0. A sub-500ms RetryAfter rounds to 0s and falls through to 'Wait before retrying.' A test for this edge case was added.

File: pkg/errors/error_test.go [hunk 4]
```diff
@@ -1,12 +1,14 @@
-package errors
+﻿package errors
```
[the remark this answers is not in the document]
Comment (danmoseley) [reply pkg/errors/error_test.go]: Fixed: removed the BOM in the follow-up commit. It was inadvertently introduced by .NET's WriteAllText with System.Text.Encoding.UTF8 (which includes a BOM by default) during a line-ending normalization step. The committed object now starts with plain 'package errors'.

**Comment by danmoseley** [comment]
@SamMorrowDrums heya, this is a very small change, any chance we could get it signed off and merged?

**Review by jluocsa — COMMENTED** [review]
Really like this — the test names are clear, the edge-case coverage (sub-second, zero time, past time, wrapped errors via `errors.As`, missing `RetryAfter`) is the kind of thing that usually only gets backfilled after a bug report. Three small things worth raising:

**1. Scope question: GraphQL rate limits.** `NewGitHubAPIErrorResponse` is the REST path. GraphQL has its own rate-limit shape (errors with `type: "RATE_LIMITED"` in the response body, not a 403 / `RateLimitError`). If a GraphQL call gets rate-limited, it currently flows through a separate handler and won't get the same clean message. Is that intentional for this PR (REST-only scope, GraphQL follow-up), or worth a parallel branch in the GraphQL error helper? Either way is fine — just want to make sure it's not an oversight.

**2. Long retry windows.** For the primary rate limit (5000/hr for authenticated requests), a 403 right after exhausting your quota can produce `Retry after 59m59s`. That's accurate but agents often retry on monotonic clocks; if they pause-and-resume across a sleep/wake, the human-readable duration is now stale. Consider including the absolute reset timestamp:

```
GitHub API rate limit exceeded. Retry after 59m59s (resets at 2025-01-15T14:30:00Z).
```

Optional / polish — but useful for long-running agent sessions.

**3. `Round(time.Second)` on a value that's already in seconds.** `rateLimitErr.Rate.Reset` comes from the `X-RateLimit-Reset` header, which is an integer epoch second. `time.Until(resetTime)` will have at most sub-second drift from the clock skew between the call and `Until`. The `.Round(time.Second)` is fine and matches the formatting intent, just noting that it's defensive rather than truly needed.

The `AbuseRateLimitError.RetryAfter` is genuinely sub-second-capable (some clients set it from `Retry-After: 0.5`), so the rounding there does real work — and the "rounds-to-zero falls back to wait message" test case is a really nice catch.

The fact that the original error is still placed on context via `addGitHubAPIErrorToContext` before the rate-limit branch returns is the right call — middleware/observability still sees the typed error.

**Comment by danmoseley** [comment]
@jluocsa thanks for the feedback. I propose to merge this as-is as it's a strict improvement standing alone, I think, and it addresses what popped in the corpus of real world agent errors. I could potentially do a follow up, but don't plan it (I moved teams since)

**Review by pachecocordovamoiseseduardo-byte — APPROVED** [review]

**Review by SamMorrowDrums — APPROVED** [review]
Clean, well-tested rate-limit/secondary-limit error messaging with retry timing. Idiomatic errors.As, handles zero/negative durations. No schema or context impact. All checks green.
